Use roughness-dependent fresnel when rendering with multiscattering

jsantell opened the pull request:
Follow up from the multiscattering handling in PBR rendering: https://github.com/mrdoob/three.js/pull/15644 https://github.com/mrdoob/three.js/pull/15673

The fresnel was not taking into account roughness, resulting in bright fresnel that is especially noticible on rough, dark materials.

(0x000, roughness:1, metalness:0) With changes (top), and r105 (bottom):

![Screenshot from 2019-05-24 18-26-43](https://user-images.githubusercontent.com/641267/58363094-08b5de00-7e54-11e9-9676-beb1ff6d88f0.png)

(0x000, roughness:1, metalness:0) With changes (top), and r105 (bottom):

![Screenshot from 2019-05-24 18-41-16](https://user-images.githubusercontent.com/641267/58363096-0e132880-7e54-11e9-81bd-71f7257e46a4.png)

jsantell committed:
Use roughness-dependent fresnel when rendering with multiscattering

WestLangley commented on src/renderers/shaders/ShaderChunk/bsdfs.glsl.js:
```diff
@@ -268,17 +279,15 @@ vec3 BRDF_Specular_GGX_Environment( const in GeometricContext geometry, const in
 void BRDF_Specular_Multiscattering_Environment( const in GeometricContext geometry, const in vec3 specularColor, const in float roughness, inout vec3 singleScatter, inout vec3 multiScatter ) {
 
-	float dotNV = saturate( dot( geometry.normal, geometry.viewDir ) );
+	float dotNV = max( 0.01, dot( geometry.normal, geometry.viewDir ) );
```
Can you please explain your reason for this redefinition?

jsantell replied:
Idea was this should never actually be 0, but that isn't true in some projections(?), and regardless doesn't change the outcome here. Will remove.

WestLangley commented on src/renderers/shaders/ShaderChunk/bsdfs.glsl.js:
```diff
@@ -268,17 +279,15 @@ vec3 BRDF_Specular_GGX_Environment( const in GeometricContext geometry, const in
+	float dotNV = max( 0.01, dot( geometry.normal, geometry.viewDir ) );
 
-	vec3 F = F_Schlick( specularColor, dotNV );
+	vec3 F = F_Schlick_RoughnessDependent( specularColor, dotNV, roughness );
```
You are passing in `dotNV` to your method, but the method as written expects `dotLH`.

jsantell replied:
The Fdez-Aguera paper uses the normal vector rather than the halfway vector. Not sure if that's required, or if this should use the halfway vector like other instances. I'll do some side by side tests.

Either way, as this is a new schlick function, the signature needs to be updated. Good catch!

jsantell replied:
Updated

jsantell commented:
@robertlong Could you give this a try if you still have that hot-fresnel in Hubs?

jsantell commented:
Addressed comments, and trying more scenarios with the changes. This could also be a preference, with prior art of PBR models (Disney's) exposing some "fresnel" parameter that isn't based on physics. Similarly, with the previous standard shader changes in r101/ #15644, any sort of change can be unexpected.

Maybe parameterizing a fresnel value by mapping directly to the roughness dampening seems like an obvious parameter to consider if anything, although I haven't given this more thought.

Side by side comparison of change; dielectrics on top, conductors bottom, roughness 0 -> 1

![three-pbr-105changes](https://user-images.githubusercontent.com/641267/58384364-dd3b0c80-7f95-11e9-9a18-c6130abd9c30.gif)

jsantell commented:
There's also the related work of potentially reverting https://github.com/mrdoob/three.js/pull/15673 -- the rationale was that it produced a dark fresnel, but that could just be that scenario (there's no dark fresnel when the torus is reflecting the lighter sky, rather than distorting the dark building).

Reverting that patch, along with this patch makes dielectrics energy-preserving as well (i.e. passing the furnace test), and may be preferable. This is what it looks like:

![three_pbr_105-dev2](https://user-images.githubusercontent.com/641267/58385789-c226c780-7faa-11e9-9aa1-a8a97aa0bd11.png)

Note that this test scene is WIP, the view angles for the objects on the side are more extreme than in the middle.

I can kick this to a new issue as well

mrdoob commented:
Have to tried comparing how the same scene looks like in babylon.js?

jsantell commented:
> Have to tried comparing how the same scene looks like in babylon.js?

Not yet, although here are some comparisons, roughness increasing rightward, metalness increasing upward in a furnace test; any visible surfaces indicate energy gain or loss:

r100
<img src="https://user-images.githubusercontent.com/641267/58438082-dc7fa480-8081-11e9-9a27-d92fa1e0a3fc.png" width="35%" />

r104, with energy compensation for metals, note the top row is invisible (#15644, #15673). Note the newly hottened fresnel, especially on rough materials. This is what's currently on dev.
<img src="https://user-images.githubusercontent.com/641267/58438206-82cbaa00-8082-11e9-864c-4b7e60c7a9ad.png" width="35%" />

Reverting #15673 makes dielectrics also energy preserving however, completely hiding the row at the bottom:
<img src="https://user-images.githubusercontent.com/641267/58438214-8c551200-8082-11e9-9d74-db9dea4daeb4.png" width="35%" />

----

Applying this patch, roughness dependent fresnel; note the reduced hotness compared to the r104 renders:
<img src="https://user-images.githubusercontent.com/641267/58438266-d6d68e80-8082-11e9-96f5-626a51073345.png" width="35%" />

Applying this patch, on top of reverting #15673:

<img src="https://user-images.githubusercontent.com/641267/58438329-261cbf00-8083-11e9-8349-eea9832db575.png" width="35%" />

----

What is the ideal outcome, or goal? Targeting another renderer (e.g. Filament, Babylon), or manufactured tests like the furnace tests above, or by arbitrary examples in the repo? I'm not even sure. At least subjectively, this patch makes rough matte objects possible again, subduing the very hot fresnel, and matching the paper's implementation

WestLangley commented:
#15673 does not look right, so I'd revert it.

This PR is correcting an error in a former PR, right? So, it has to be fixed -- unless you argue that the change is not perceptible (e.g., your last image above).

You may be seeing what happens when we cobble together techniques from multiple sources that do not fit together as a coherent whole.

jsantell commented:
> #15673 does not look right, so I'd revert it.

The [original rational](https://github.com/mrdoob/three.js/pull/15644#issuecomment-458719044) was a dark fresnel in some cases, but I think that may have been due to some specific examples (and think they were correctly reflecting dark environments). // @mrdoob 

> 
> This PR is correcting an error in a former PR, right? So, it has to be fixed -- unless you argue that the change is not perceptible (e.g., your last image above).

This PR is correcting the original energy compensation one by applying roughness-dependent fresnel. This patch, and the reverting are two additional changes -- when applying *both* (last image), the bottom row is now invisible and overall darkening for non-dielectric/metal materials (due to revert), and additionally, there is less fresnel glow (due to this patch). Right now, without either patch, the fresnel is very hot (second picture). Either of these would reduce that delta, but I believe both would be the most consistently correct in terms of furnace test at least. Handling the inbetween (0<metalness<1) values can be done with Filament's technique + LUT IIRC, but everything would be consistent with the Fdez paper with both of these.

> You may be seeing what happens when we cobble together techniques from multiple sources that do not fit together as a coherent whole.

I was thinking the same thing, hence why I'm mentioning reverting #15673 at all, since they're all related.

mrdoob commented:
Reverted #15673: https://github.com/mrdoob/three.js/commit/1128d601c1b57eeee49678216bbb5dcabf8efc98

mrdoob commented:
Thanks!

mrdoob commented:
Yeah, that fresnel was a bit too much...

Before:

<img width="1220" alt="Screen Shot 2019-05-27 at 6 34 36 PM" src="https://user-images.githubusercontent.com/97088/58444900-11eeb700-80af-11e9-87f2-ead9f22bcc09.png">

After:

<img width="1220" alt="Screen Shot 2019-05-27 at 6 34 49 PM" src="https://user-images.githubusercontent.com/97088/58444907-17e49800-80af-11e9-9d41-43a66cc8a87d.png">
